streams/shared/src/main/scala/zio/stream/ZTransducer.scala

```diff
@@ -124,10 +124,38 @@ object ZTransducer {
     }
 
   /**
-   * Creates a transducer accumulating incoming values into lists of maximum size `n`.
+   * Creates a transducer accumulating incoming values into chunks of maximum size `n`.
    */
-  def collectAllN[I](n: Long): ZTransducer[Any, Nothing, I, List[I]] =
-    foldUntil[I, List[I]](Nil, n)((list, element) => element :: list).map(_.reverse).filter(_.nonEmpty)
+  def collectAllN[I](n: Long): ZTransducer[Any, Nothing, I, Chunk[I]] =
+    ZTransducer {
+
+      def go(in: Chunk[I], builder: ChunkBuilder[I], size: Long): (ChunkBuilder[Chunk[I]], ChunkBuilder[I], Long) =
+        in.foldLeft[(ChunkBuilder[Chunk[I]], ChunkBuilder[I], Long)]((ChunkBuilder.make(), builder, size)) {
+          case ((outBuilder, builder, size), i) =>
+            if (size + 1 == n) (outBuilder += (builder += i).result(), ChunkBuilder.make(n.toInt), 0)
+            else (outBuilder, builder += i, size + 1)
+        }
+
+      ZRef.makeManaged[(ChunkBuilder[I], Long)]((ChunkBuilder.make(n.toInt), 0)).map { stateRef =>
+        {
+          case None =>
+            stateRef
+              .getAndSet((ChunkBuilder.make(n.toInt), 0))
+              .map {
+                case (chunkBuilder, chunkSize) =>
+                  if (chunkSize == 0) Chunk.empty
+                  else Chunk(chunkBuilder.result())
+              }
+
+          case Some(in) =>
+            stateRef.modify {
```

Comment:
`ChunkBuilder` could be reused after emitting a chunk, if I'm not wrong

Reply:
It could be reused and I almost went for it, but after giving it a thought I'd argue that it's reusability is implementation details as ChunkBuilder is not [`ReusableBuilder`](https://www.scala-lang.org/api/current/scala/collection/mutable/ReusableBuilder.html) as of right now.

Reply:
`ChunkBuilder` has a public  `clear` method, so yes